### PR TITLE
Overhaul Python common package build

### PR DESCRIPTION
**Pull request by marcelamelara**
This PR overhauls how we build and install the python library:

- [x] Switches the build for the SWIG Python bindings to use the updated cmake build infra
- [x] Consolidates client and core python packages
- [x] Builds a Python wheel for the common/client and sservice
~~- [ ] Write pdo.common.crypto wrapper~~
- [x] Documents updated python package structure

### CODE REVIEW

**Comment by cmickeyb** [comment]
This looks like a great start! I'll have to spend more time with the CMakefile to understand all the ins & outs of what you're doing there. 

One other note... should we consider packaging the SWIG files separately? that is, move them out of the python directory tree into one of their own and keep the python tree strictly python? (and maybe combine it with the client tree)

**Comment by cmickeyb** [comment]
One thing I would really like to see from this... can we get to where we can build the packages without the virtual environment in place? Clearly we need the dependencies to run, but do we need them to build? This is part of pushing for a clean separation between the build, install, configure and run phases.

**Comment by marcelamelara** [comment]
> can we get to where we can build the packages without the virtual environment in place? Clearly we need the dependencies to run, but do we need them to build? 

I'll look into this, because agreed, it would be nice to separate out the different phases more cleanly.

File: docs/figures/contract-invoke.excalidraw [hunk 1]
```diff
@@ -0,0 +1,1412 @@
...
+      "originalText": "end user",
+      "lineHeight": 1.25
+    }
+  ],
+  "appState": {
+    "gridSize": null,
+    "viewBackgroundColor": "#ffffff"
+  },
+  "files": {}
+}
```
Comment (cmickeyb) [inline docs/figures/contract-invoke.excalidraw:1412]: is there a particular reason to include this? have you checked license implications?

Comment (cmickeyb) [reply docs/figures/contract-invoke.excalidraw:1412]: Is Excalidraw a common tool? I've never heard of it. Any reason not to make them something more commonly used (google draw?). Or maybe i'm just way behind the times.

Comment (marcelamelara) [reply docs/figures/contract-invoke.excalidraw:1412]: I think excalidraw is actually being used under the hood by other whiteboard applications, I just realized recently that there's a separate in-browser service as well.

**Comment by marcelamelara** [comment]
> This might make it possible to remove all ledger-specific code from the base python packages.

All components use the ledger APIs, though, don't they? I'd rather keep as much of the python code in one place as possible. It makes it a bit easier to track, IMO, and also makes the python package build more straight-forward. That said, I am in favor of reproducing the `ledgers/ccf` directory structure within `python/pdo`.

**Comment by cmickeyb** [comment]
> > This might make it possible to remove all ledger-specific code from the base python packages.
> 
> All components use the ledger APIs, though, don't they? I'd rather keep as much of the python code in one place as possible. It makes it a bit easier to track, IMO, and also makes the python package build more straight-forward. That said, I am in favor of mirroring the `ledgers/ccf` directory structure within `python/pdo`.

The only ledger specific code we have right now is the submitter and a few of the key functions. Those can be pulled into the package (setup includes the designated directory in the ledger/$PDO_LEDGER_TYPE/python directory. There is a super class for the submitter that stays in python...

**Comment by marcelamelara** [comment]
Maybe I'm missing something, but why do we want to build and configure CCF separately? I understand doing this for the TP, though the submitter is used by several other components besides the client.

I can move the CCF-specific submitter to the ccf directory since it's just one file, but in general, I would like to avoid/minimize having more python code spread across a bunch of different directories. Right now we have python code mixed with cpp code in practically every directory, and I think the dependencies will be clearer, and also easier to maintain, if we separate the cpp layers from the python layers.

**Comment by cmickeyb** [comment]
> Maybe I'm missing something, but why do we want to build and configure CCF separately? I understand doing this for the TP, though the submitter is used by several other components besides the client.
> 
> I can move the CCF-specific submitter to the ccf directory since it's just one file, but in general, I would like to avoid/minimize having more python code spread across a bunch of different directories. Right now we have python code mixed with cpp code in practically every directory, and I think the dependencies will be clearer, and also easier to maintain, if we separate the cpp layers from the python layers.

i don't want to build it separately. I want the python to pull the packages from PDO_SOURCE_ROOT/ledgers/ccf/python instead. That way all of the ccf code is in one place rather than scattered throughout PDO_SOURCE_ROOT/python.

**Comment by cmickeyb** [comment]
>Right now we have python code mixed with cpp code in practically every directory, and I think the dependencies will be clearer, and also easier to maintain, if we separate the cpp layers from the python layers.

This is precisely because we have separate packages. We do NOT want the eservice or pservice code being packaged with the client package because it would pull in all the sgx dependencies. We should be building something like:
* eservice package
* pservice package
* sservice package
* client package
* service support package

we may also want a "ledger" package. but i'm ok rolling this into the client/service support packages

**Comment by marcelamelara** [comment]
@cmickeyb We currently ship the contents of `$PDO_HOME/opt/pdo/bin` and `$PDO_HOME/opt/pdo/etc` with pdo.common (see python/setup.py data_files parameter to setup). I don't think we need either of those anymore? Especially the contents of /etc, which include ledger keys, seem rather problematic to distribute as part of pdo.common...

**Comment by cmickeyb** [comment]
> @cmickeyb We currently ship the contents of `$PDO_HOME/opt/pdo/bin` and `$PDO_HOME/opt/pdo/etc` with pdo.common (see python/setup.py data_files parameter to setup). I don't think we need either of those anymore? Especially the contents of /etc, which include ledger keys, seem rather problematic to distribute as part of pdo.common...

I think we can drop the etc & bin directories. Nothing goes in them right now.

**Comment by marcelamelara** [comment]
Closing this to break up into multiple PRs.

**Pull request closed without merging**